File: ERP/xe-toolkit/src/java/XE/com/capmet/tools/XE/programs/IoStat.java
/****************************************************************************
**
** Copyright (c) 2006-2007 Captive Metrics Software Corporation.
**                    All rights reserved.
**
** This file is part of the XE programs for The XE Toolkit.
**
** This file may be used under the terms of the GNU General Public
** License version 2.0 as published by the Free Software Foundation
** and appearing in the file LICENSE-GPL.txt contained within the
** same package as this file. This software is subject to a
** dual-licensing mechanism, the details of which are outlined in
** file LICENSE-DUAL.txt, also contained within this package. Be sure
** to use the correct license for your needs. To view the commercial
** license, read LICENSE-COMMERCIAL.txt also contained within this
** package.
**
** If you do not have access to these files or are unsure which license
** is appropriate for your use, please contact the sales department at
** sales@captivemetrics.com.
**
** This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
** WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
**
****************************************************************************/

package com.capmet.tools.XE.programs;

import java.util.Map;
import com.capmet.tools.XE.GetOpt;
import com.capmet.tools.XE.XEMessages;
import com.capmet.metrics.cm.CmIoStat;
import com.capmet.metrics.cm.CmIoMetrics;
import com.capmet.metrics.cm.CmMounts;

/**
 * An application for monitoring the I/O entities on a host.<p>
 *
 * Use: java IoStat [-dN] [-n disk] [-h remote-host] [interval [iterations]]<br>
 *      -d - show physical disks only<br>
 *      -N - show network disks<br>
 *      -n disk - show only named disk<br>
 *      -h remote-host - monitor remote host
 */
public class IoStat
{
    private final int HEADER_INTERVAL = 20;
    private int lineCount = 0;

    private CmMounts diskMap;

    /**
     * The main method of the IoStat application.
     * @param args Arguments passed from the environment
     */
    public static void main(String[] args)
    {
        new IoStat(args);
    }

    // usage blurb
    private void usage()
    {
        System.out.println(XEMessages.msgs.getString("IoStat_Usage"));
    }

    /**
     * IoStat constructor.
     * @param args Arguments passed from the environment
     */
    public IoStat(String[] args)
    {
        try {
            boolean showDisks = false;
            boolean showNetDisks = false;
            String namedDisk = null;
            String remoteHost = null;
            int iterations = 1;
            int interval = 1;
            GetOpt opts = new GetOpt(args, "dn:Nh:");
            int c;

            opts.optErr = false;
            while((c = opts.getopt()) != opts.optEOF) {
                switch(c) {
                case 'd':
                    showDisks = true;
                    break;
                case 'n':
                    namedDisk = opts.getOptArg();
                    break;
                case 'N':
                    showNetDisks = true;
                    break;
                case 'h':
                    remoteHost = opts.getOptArg();
                    break;
                default:
                    usage();
                    return;
                }
            }

            int optind = opts.getOptIndex();
            int argCount = args.length - optind;
            switch(argCount) {
            case 0:
                break;
            case 1:
                try {
                    interval = Integer.parseInt(args[optind]);
                    iterations = 0;
                } catch(NumberFormatException e) {
                    usage();
                    return;
                }
                break;
            case 2:
                try {
                    interval = Integer.parseInt(args[optind++]);
                    iterations = Integer.parseInt(args[optind]);
                } catch(NumberFormatException e) {
                    usage();
                    return;
                }
                break;
            default:
                usage();
                return;
            }

            if (interval == 0)
                interval = 1;
            if (iterations == 0)
                iterations = Integer.MAX_VALUE;

            CmIoStat io = new CmIoStat(remoteHost);
            diskMap = new CmMounts(remoteHost);

            do {
                if (lineCount == 0) {
                    System.out.println(
                        "device           %util   qlen (rio:kB rd  wio:kB wr" +
                        "  ios:kB /sec) svctim (ms)");
                    lineCount = HEADER_INTERVAL;
                }

                for(Map.Entry<String, CmIoMetrics> e : io.entrySet())
                {
                    String name = e.getKey();
                    CmIoMetrics iom = e.getValue();

                    boolean printit = false;
                    boolean mapit = iom.diskType.equals("logical");

                    if (mapit)
                        name = diskMap.toMountPoint(name);

                    // if a disk was named and this is the one
                    if (namedDisk != null) {
                        if (namedDisk.equals(name))
                            printit = true;
                    } else {
                        // print those requested
                        // skip the _Total disk
                        if (name.equals("_Total"))
                            printit = false;
                        else if (iom.diskType.equals("logical"))
                            printit = true;
                        else if (iom.diskType.equals("physical"))
                        {
                            printit = showDisks;
                            int space = name.indexOf(' ');
                            if (space != -1)
                                name = name.substring(0, space);
                        }
                        else if (iom.diskType.equals("network"))
                            printit = showNetDisks;
                    }

                    if (printit) {
                        System.out.printf(
                            "%-14.14s %6.2f%% %6.2f  %3d:%-6d %3d:%-6d %3d:%-6d %8.1f\n",
                                name, iom.utilizationPercent.doubleValue(),
                                iom.averageQueueLength.doubleValue(),
                                iom.readOperations.intValue(), iom.kilobytesRead.intValue(),
                                iom.writeOperations.intValue(), iom.kilobytesWritten.intValue(),
                                iom.readOperations.intValue() + iom.writeOperations.intValue(),
                                iom.kilobytesRead.intValue() + iom.kilobytesWritten.intValue(),
                                iom.averageAccessTime.doubleValue());
                        lineCount--;
                    }
                }

                if (--iterations > 0) {
                    Thread.sleep(interval * 1000);
                    io.refresh();
                }
            } while(iterations > 0);
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
}

File: ERP/xe-toolkit/src/mdr/tools/iostat.cpp
/****************************************************************************
**
** Copyright (c) 2006-2007 Captive Metrics Software Corporation.
**                    All rights reserved.
**
** This file is part of the mdr tools of Captivity.
**
** This file may be used under the terms of the GNU General Public
** License version 2.0 as published by the Free Software Foundation
** and appearing in the file LICENSE-GPL.txt contained within the
** same package as this file. This software is subject to a
** dual-licensing mechanism, the details of which are outlined in
** file LICENSE-DUAL.txt, also contained within this package. Be sure
** to use the correct license for your needs. To view the commercial
** license, read LICENSE-COMMERCIAL.txt also contained within this
** package.
**
** If you do not have access to these files or are unsure which license
** is appropriate for your use, please contact the sales department at
** sales@captivemetrics.com.
**
** This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
** WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
**
****************************************************************************/


#include <iostream>
#include <unistd.h>
#include <sys/ioctl.h>
#ifdef _SOLARIS
# include <sys/termios.h>
#endif
#include <mdr_iostat.h>
#include <mdr_mnt.h>
#include <mdr_exception.h>

using namespace mdr;

static inline void
header (void)
{
  puts ("device           %util   qlen (rio:kB rd  wio:kB wr"
        "  ios:kB /sec) svctim (ms)");
}

int
main (int argc, char *argv[])
{
  try
  {
    md_node_t n;
    md_can_t *cp;
    std::string name;
    char *wanted_disk = 0;
    int c;
    bool printit;
    int interval = 1;
    int iterations = 1;
    bool show_disks = false;
    bool chatty_header = false;
    bool net_disk = false;
    int line_count = 20;
    int lines_til_header = line_count;
    struct winsize winsz;
    const char *remote_host = 0;
    bool do_support_matrix = false;

    opterr = 0;
    while ((c = getopt (argc, argv, "cdNn:sh:")) != -1)
      {
        switch (c)
          {
          case 'c':
            chatty_header = true;
            break;
          case 'd':
            show_disks = true;
            break;
          case 'n':
            wanted_disk = optarg;
            break;
          case 'N':
            net_disk = true;
            break;
          case 's':
            do_support_matrix = true;
            break;
          case 'h':
            remote_host = optarg;
            break;
          default:
            fprintf(stderr, "Use: %s [-cdNs] [-n disk] [-h remote_host] "
                            "[interval [iterations]]\n",
                     argv[0]);
            return 1;
          }
      }

    mdr_iostat_t io(remote_host);
    mdr_mnt_t mnt(remote_host);

    if (do_support_matrix) {
      mdr_t::mdr_support_matrix(io, std::cout);
      return 0;
    }

    if (argv[optind])
      {
        interval = atoi (argv[optind]);
        if (argv[optind + 1])
          iterations = atoi (argv[optind + 1]);
        else
          iterations = 0;
      }

    if (!chatty_header)
      {
        if (ioctl (1, TIOCGWINSZ, &winsz) != -1)
          line_count = winsz.ws_row - 3;
        lines_til_header = line_count;
        header ();
      }

    do
      {
        if (chatty_header)
          header ();
        for (n = io.begin (); n != io.end (); n++)
          {
            cp = &n->second;
            name = n->first;
            printit = false;
            mmnt_node_t mn = mnt.find(name);
            if (mn != mnt.end())
              name = mn->second.mmnt_mount_point;

            // if wanted_disk and this is it
            if (wanted_disk)
              {
                if (name == wanted_disk)
                  printit = true;
              }
            else
              {
                if (cp->md_type == "logical")
                  {
                    if (!show_disks)
                      printit = true;
                  }
                else if (cp->md_type == "physical")
                  {
                    if (show_disks)
                      printit = true;
                  }
                else if (cp->md_type == "network")
                  {
                    if (net_disk)
                      printit = true;
                  }
              }

            if (printit) {
              printf
                ("%-14.14s %6.2f%% %6.2f  %3u:%-6u %3u:%-6u %3u:%-6u %8.1f\n",
                 name.c_str(), cp->md_util_percent,
                 cp->md_avg_queue_length, cp->md_reads, cp->md_read_K,
                 cp->md_writes, cp->md_written_K,
                 cp->md_io_operations,
                 cp->md_read_K + cp->md_written_K,
                 cp->md_service_time);

              if (!chatty_header && (--lines_til_header == 0))
                {
                  lines_til_header = line_count;
                  header ();
                }
            }
          }
        if (iterations != 1)
          {
            sleep (interval);
            io.mdr_refresh();
          }
      }
    while (--iterations);
  }
  catch (std::exception & e)
  {
    std::cout << argv[0] << ": " << e.what () << std::endl;
  }

  return 0;
}
